packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx

```diff
@@ -18,21 +26,115 @@ interface Props {
   setHideToolbarControls?: (disabled: boolean) => void;
 }
 
-export const HtmlView: FC<Props> = ({ document, setHideToolbarControls, setLoading }) => {
+export const canRenderHtmlView = (document?: QueryResult): boolean => !!get(document, 'html');
+
+const SANITIZE_CONFIG = {
+  ADD_TAGS: ['bbox'],
+  ADD_ATTR: [
+    // bbox
+    'page'
+  ],
+  WHOLE_DOCUMENT: true
+};
+
+const base = `${settings.prefix}--html`;
+
+export const HtmlView: FC<Props> = ({
+  document,
+  highlight,
+  setHideToolbarControls,
+  setLoading
+}) => {
+  const contentRef = useRef<HTMLDivElement>(null);
+  const highlightRef = useRef<HTMLDivElement>(null);
+
   if (setHideToolbarControls) {
     setHideToolbarControls(true);
   }
 
-  const html = document ? DOMPurify.sanitize(get(document, 'html')) : '';
+  const [html, setHtml] = useState<string | null>(null);
+  useEffect(() => {
+    if (document) {
+      const docHtml = document.html;
+      if (docHtml && highlight) {
+        // "process" the original HTML string in order to find offsets for DOM nodes
+        const process = async (): Promise<void> => {
+          const processedDoc = await processDoc({ ...document, docHtml }, { sections: true });
+          const fullHtml = processedDoc.sections
+            ? processedDoc.sections.map(section => section.html).join('')
+            : '';
+
+          // set sanitized HTML (removing scripts, etc)
+          setHtml(`
+          <div>
+            <style>${processedDoc.styles}</style>
+            ${DOMPurify.sanitize(fullHtml, SANITIZE_CONFIG)}
+          </div>`);
+        };
+        process();
+      } else {
+        // if no highlight, then no need to "process"
+        setHtml(docHtml ? DOMPurify.sanitize(docHtml, SANITIZE_CONFIG) : '');
+      }
+    }
+
+    setLoading(false);
+  }, [document, highlight, setLoading]);
```

Comment:
i think the `useEffect` here will re-run a bunch for these non-primitive, non-react-generated input parameters. not sure if that is a concern for performance as I haven't tested it locally. one option is to use the `useDeepCompareEffect` which uses lodash's `isEqual` method on complex object types

regardless, this is a performance optimization only to revisit if we see things being slow, etc.

Reply:
So far, haven't seen perf issues.